### PR TITLE
chore: clean encrypted training and SGDClassifier files

### PR DESCRIPTION
**Pull request by RomanBredehoft**
I was starting to work on https://github.com/zama-ai/concrete-ml-internal/issues/4374 but realized the SGDClassifier could benefit from a bit of cleaning

also wanted to fix https://github.com/zama-ai/concrete-ml-internal/issues/4184 but it's actually not that trivial so we'll see later if it's really needed

### CODE REVIEW

File: src/concrete/ml/sklearn/linear_model.py [hunk 1]
```diff
@@ -398,6 +400,8 @@ def _fit_encrypted(
                 optimization. Default to None.
             intercept_init (Optional[numpy.ndarray]): The initial intercept to warm-start the
                 optimization. Default to None.
+            is_partial_fit (bool): Indicates if this fit represents a partial fit. A partial fit is
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/linear_model.py:403]: basically I removed `_encrypted_fit_one_step` (made for `partial_fit`) and added this new param in `_fit_encrypted` to handle the few differences with between `fit` and `partial_fit`

File: src/concrete/ml/sklearn/linear_model.py [hunk 2]
```diff
@@ -418,90 +422,115 @@ def _fit_encrypted(
...
 
-        # Initialize the weight values
+        # Initialize the weight values with the given ones if some are provided
         if coef_init is not None:
             weights = coef_init
-        elif self.warm_start and self._weights_encrypted_fit is not None:
+
+        # Else, if warm start is activated or this is a partial fit, use some already computed
+        # weight values have if there are some
+        elif (self.warm_start or is_partial_fit) and self._weights_encrypted_fit is not None:
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/linear_model.py:484]: this is more or less a fix : before, we were asking users to set `warm_start` to True in order to benefit from `partial_fit`, but that's not the case in sklearn (https://scikit-learn.org/stable/glossary.html#term-partial_fit)

File: src/concrete/ml/sklearn/linear_model.py [hunk 3]
```diff
@@ -663,132 +709,6 @@ def fit(  # type: ignore[override]
             sample_weight=sample_weight,
         )
 
-    # pylint: disable-next=too-many-branches,too-many-statements
-    def _fit_encrypted_one_step(self, X, y, fhe):
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/linear_model.py:667]: as I mentioned above, this is no longer needed (and it removes a lot of duplicated code)

File: tests/sklearn/test_fhe_training.py [hunk 4]
```diff
@@ -293,259 +302,236 @@ def test_clear_fit(
     model.predict_proba(x, fhe="simulate")
 
 
-# pylint: disable=too-many-statements,protected-access,too-many-locals
-@pytest.mark.parametrize("fit_intercept", [True, False])
-@pytest.mark.parametrize("label_offset", [0, 1])
-@pytest.mark.parametrize("n_bits, max_iter, parameter_min_max", [pytest.param(7, 30, 1.0)])
-def test_encrypted_fit_coherence(
-    fit_intercept, label_offset, n_bits, max_iter, parameter_min_max, check_accuracy
+def check_encrypted_fit(
```
Comment (RomanBredehoft) [inline tests/sklearn/test_fhe_training.py:305]: I refactored the test by creating this function

File: tests/sklearn/test_fhe_training.py [hunk 5]
```diff
@@ -293,259 +302,236 @@ def test_clear_fit(
...
+    assert array_allclose_and_same_shape(y_pred_proba_disable, y_pred_proba_mixed)
+    assert array_allclose_and_same_shape(y_pred_class_disable, y_pred_class_mixed)
+
+    # Define warn start parameters
+    warm_start_fit_kwargs = {
+        "coef_init": weights_disable,
+        "intercept_init": bias_disable,
+    }
+
+    # Because the model's `random_number_generator` will be re-initialized with the same random
```
Comment (RomanBredehoft) [inline tests/sklearn/test_fhe_training.py:523]: I tried to add checking values coherence at the end but unfortunately the fact that we re-initialize the rng instance at each model init with the exact same random_state more or less prevents us to do so

File: src/concrete/ml/sklearn/linear_model.py [hunk 6]
```diff
@@ -361,19 +360,21 @@ def _get_training_quantized_module(
...
         )
         end = time.time()
+
         if self.verbose:
             print(f"Compilation took {end - start:.4f} seconds.")
 
         return training_quantized_module
 
-    # pylint: disable-next=too-many-branches, too-many-statements
+    # pylint: disable-next=too-many-branches, too-many-statements, too-many-locals
```
Comment (fd0r) [inline src/concrete/ml/sklearn/linear_model.py:369]: 🙈

File: src/concrete/ml/sklearn/linear_model.py (hunk 1 above)
Comment (fd0r) [reply src/concrete/ml/sklearn/linear_model.py:402]: That's indeed probably a better solution than the 2 method one we had before!

File: src/concrete/ml/sklearn/linear_model.py (hunk 2 above)
Comment (fd0r) [reply src/concrete/ml/sklearn/linear_model.py:483]: Good catch!

**Review by fd0r — APPROVED** [review]

File: src/concrete/ml/sklearn/linear_model.py (hunk 6 above)
Comment (RomanBredehoft) [reply src/concrete/ml/sklearn/linear_model.py:369]: 🙉

File: tests/sklearn/test_fhe_training.py (hunk 5 above)
Comment (RomanBredehoft) [reply tests/sklearn/test_fhe_training.py]: just changed and made passing the RNG object between calls possible indeed, the test is green , thanks !

**Review by fd0r — APPROVED** [review]

**Pull request merged**